lib/fast-path.ts

```diff
@@ -347,205 +344,208 @@ FPp.needsParens = function (assumeExpressionContext) {
     return false;
   }
 
-  if (
-    parent.type === "ParenthesizedExpression" ||
-    (node.extra && node.extra.parenthesized)
-  ) {
+  if (parent && parent.type === "ParenthesizedExpression") {
     return false;
   }
 
-  switch (node.type) {
-    case "UnaryExpression":
-    case "SpreadElement":
-    case "SpreadProperty":
-      return (
-        parent.type === "MemberExpression" &&
-        name === "object" &&
-        parent.object === node
-      );
-
-    case "BinaryExpression":
-    case "LogicalExpression":
-      switch (parent.type) {
-        case "CallExpression":
-          return name === "callee" && parent.callee === node;
-
-        case "UnaryExpression":
-        case "SpreadElement":
-        case "SpreadProperty":
-          return true;
+  if (node.extra && node.extra.parenthesized) {
+    return true;
+  }
```

Comment:
The problem with short circuiting to `false` as we did previously is that a node may still really **need** parens (that is, non-decoratively). I opted to make this code responsible for decorative parens which as far as I can tell works fine, though it isn't apparent to me yet if there is some reason the code was written the way it was. The test suite seems to indicate that there is not, or at least is no longer.